core/src/main/java/org/apache/iceberg/util/TableScanUtil.java

```diff
@@ -128,6 +136,61 @@ public static <T extends ScanTask> CloseableIterable<ScanTaskGroup<T>> planTaskG
         combinedTasks -> new BaseScanTaskGroup<>(mergeTasks(combinedTasks)));
   }
 
+  @SuppressWarnings("unchecked")
+  public static <T extends PartitionScanTask> List<ScanTaskGroup<T>> planTaskGroups(
+      List<T> tasks,
+      long splitSize,
+      int lookback,
+      long openFileCost,
+      Types.StructType projectedPartitionType) {
+
+    Preconditions.checkArgument(splitSize > 0, "Invalid split size (negative or 0): %s", splitSize);
+    Preconditions.checkArgument(
+        lookback > 0, "Invalid split planning lookback (negative or 0): %s", lookback);
+    Preconditions.checkArgument(
+        openFileCost >= 0, "Invalid file open cost (negative): %s", openFileCost);
+
+    Function<T, Long> weightFunc =
+        task -> Math.max(task.sizeBytes(), task.filesCount() * openFileCost);
+
+    Map<Integer, StructProjection> projectionsBySpec = Maps.newHashMap();
+
+    // Group tasks by their partition values
+    StructLikeMap<List<T>> tasksByPartition = StructLikeMap.create(projectedPartitionType);
+
+    for (T task : tasks) {
+      PartitionSpec spec = task.spec();
+      StructProjection projectedStruct =
+          projectionsBySpec.computeIfAbsent(
+              spec.specId(),
+              specId -> StructProjection.create(spec.partitionType(), projectedPartitionType));
+      List<T> taskList =
+          tasksByPartition.computeIfAbsent(
+              projectedStruct.copyFor(task.partition()), k -> Lists.newArrayList());
+      if (task instanceof SplittableScanTask<?>) {
+        ((SplittableScanTask<? extends T>) task).split(splitSize).forEach(taskList::add);
+      } else {
+        taskList.add(task);
+      }
+    }
+
+    // Now apply task combining within each partition
+    return tasksByPartition.values().stream()
+        .flatMap(ts -> toTaskGroupStream(ts, splitSize, lookback, weightFunc))
+        .collect(Collectors.toList());
+  }
+
+  private static <T extends ScanTask> Stream<ScanTaskGroup<T>> toTaskGroupStream(
+      Iterable<T> tasks, long splitSize, int lookback, Function<T, Long> weightFunc) {
+    CloseableIterable<ScanTaskGroup<T>> taskGroups =
```

Comment:
After a closer look, do we need `CloseableIterable`? I don't think there is anything to close. What about using plain iterables instead of the stream then?

```
private static <T extends ScanTask> Iterable<ScanTaskGroup<T>> toTaskGroupIterable(
    Iterable<T> tasks, long splitSize, int lookback, Function<T, Long> weightFunc) {

  return Iterables.transform(
      new BinPacking.PackingIterable<>(tasks, splitSize, lookback, weightFunc, true),
      combinedTasks -> new BaseScanTaskGroup<>(mergeTasks(combinedTasks)));
}
```

We could call it like this:

```
return FluentIterable.from(tasksByPartition.values())
    .transformAndConcat(ts -> toTaskGroupIterable(ts, splitSize, lookback, weightFunc))
    .toList();
```

I feel `transformAndConcat` in `FluentIterable` would be best.

Reply:
Good idea. Let me change it.